features/collection/collectors/src/main/java/org/opennms/netmgt/collectd/JMXCollector.java

```diff
@@ -258,34 +265,68 @@ public CollectionSet collect(CollectionAgent agent, Map<String, Object> map) {
                 @Override
-                public void process(JmxAttributeSample attributeSample) {
-                    final String objectName = attributeSample.getMbean().getObjectname();
+                public void process(JmxAttributeSample attributeSample, ObjectName objectName) {
+                    final String mbeanObjectName = attributeSample.getMbean().getObjectname();
                     final String attributeName = attributeSample.getCollectedAttribute().getName();
-
-                    final String dsKey = objectName + "|" + attributeName;
+                    final String dsKey = mbeanObjectName + "|" + attributeName;
                     final JMXDataSource ds = nodeInfo.getDsMap().get(dsKey);
                     if (ds == null) {
                         LOG.info("Could not find datasource for {}. Skipping.", dsKey);
                         return;
                     }
-                    addNumericAttributeToCollectionSet(ds, attributeSample);
+
+                    String keyField = attributeSample.getMbean().getKeyfield();
+                    if (keyField != null) {
+                        final String parsedObjectName = fixGroupName(objectName.getCanonicalName());
+                        final Resource resource = new DeferredGenericTypeResource(nodeResource, keyField,
+                                parsedObjectName);
+                        addNumericAttributeToCollectionSet(ds, attributeSample, resource);
+                        addStringAttributesToCollectionSet(ds, attributeSample, resource, objectName);
+                    } else {
+                        addNumericAttributeToCollectionSet(ds, attributeSample, nodeResource);
+                    }
                 }
 
                 @Override
-                public void process(JmxCompositeSample compositeSample) {
-                    final String objectName = compositeSample.getMbean().getObjectname();
+                public void process(JmxCompositeSample compositeSample, ObjectName objectName) {
+                    final String mbeanObjectName = compositeSample.getMbean().getObjectname();
                     final String attributeName = compositeSample.getCollectedAttribute().getName();
 
-                    final String dsKey = objectName + "|" + attributeName + "|" + compositeSample.getCompositeKey();
+                    final String dsKey = mbeanObjectName + "|" + attributeName + "|"
+                            + compositeSample.getCompositeKey();
                     final JMXDataSource ds = nodeInfo.getDsMap().get(dsKey);
                     if (ds == null) {
                         LOG.info("Could not find datasource for {}. Skipping.", dsKey);
                         return;
                     }
-                    addNumericAttributeToCollectionSet(ds, compositeSample);
+                    String keyField = compositeSample.getMbean().getKeyfield();
```

Comment:
Using the `keyfield` may be a little confusing here since we normally refer to this as the `resource-type` in other collection definitions.